File: A2/Petri-Net/transition.go
package main

import {
  "os"
}

type transition struct {
  M int
  N int
  inputs [M]chan int //M input channels, cannot be 0
  outputs [N]chan int //N output channels, cannot be 0
}

//make M input channels (buffer size 1) & N output channels (buffer size 1)
func New(M int, N int) {
  var inputs [M]chan int
  for i := range inputs {
     inputs[i] = make(chan int, 1)
  }

  var outputs [N]chan int
  for j := range outputs {
     outputs[i] = make(chan int, 1)
  }

  t = transition {M, N, inputs, outputs}
  return t
}

//check if all M inputs are full
func (t transition) Eligible() {
  eligible := true
  for i := range t.inputs {
    if len(t.inputs[i]) != cap(t.inputs[i]){  //if any input channel is not full, transition is not eligible to fire
      eligible = false
    }
  }
  return eligible
}

//consume 1 token from each input & produce 1 token to each output
func (t transition) Fire() {
  for i := range t.inputs {
    <- t.inputs[i]
  }

  token := 1  //token = int with value 1
  for j := range t.outputs {
    t.outputs[i] <- token
  }
}

func (t transition) Run() {
  for {
    if t.Eligible(){
      t.Fire()
    }
  }
}

File: A2/Petri-Net/transition.py
from threading import Thread
import Queue

class Transition(Thread)

    def __init__(self, M, N):
        super(Transition, self).__init__()
        self.M = M
        self.N = N
        self.inputs = [Queue.Queue(1)] * M
        self.outputs = [Queue.Queue(1)] * N

    def eligible(self):
        elig = True
        for queue in inputs:
            if queue.empty():
                elig = False
        return elig

    def fire(self):
        for queue in inputs:
            queue.get()
        for queue in outputs:
            queue.put(1)

    def run(self):
        while True:
            if self.eligible():
                self.fire()
